src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,799 @@
+#include "thread/network_data_leader.hpp"
+#include "thread/network_data_local.hpp"
+#include "thread/network_data_notifier.hpp"
+
+namespace ot {
+
+namespace BorderRouter {
+
+RoutingManager::RoutingManager(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mIsRunning(false)
+    , mInfraIfIndex(0)
+    , mRouterAdvertisementTimer(aInstance, HandleRouterAdvertisementTimer, this)
+    , mRouterAdvertisementCount(0)
+    , mRouterSolicitTimer(aInstance, HandleRouterSolicitTimer, this)
+    , mRouterSolicitCount(0)
+    , mDiscoveredOnLinkPrefixInvalidTimer(aInstance, HandleDiscoveredOnLinkPrefixInvalidTimer, this)
+{
+    mLocalOmrPrefix.Clear();
+    memset(mAdvertisedOmrPrefixes, 0, sizeof(mAdvertisedOmrPrefixes));
+    mAdvertisedOmrPrefixNum = 0;
+
+    mLocalOnLinkPrefix.Clear();
+    mAdvertisedOnLinkPrefix = nullptr;
+    mDiscoveredOnLinkPrefix.Clear();
+}
+
+otError RoutingManager::Init(uint32_t aInfraIfIndex)
+{
+    otError error;
+
+    OT_ASSERT(!IsInitialized() && !Get<Mle::MleRouter>().IsAttached());
+    VerifyOrExit(aInfraIfIndex > 0, error = OT_ERROR_INVALID_ARGS);
+
+    SuccessOrExit(error = LoadOrGenerateRandomOmrPrefix());
+    SuccessOrExit(error = LoadOrGenerateRandomOnLinkPrefix());
+
+    mInfraIfIndex = aInfraIfIndex;
+
+exit:
+    return error;
+}
+
+otError RoutingManager::LoadOrGenerateRandomOmrPrefix(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    if (Get<Settings>().ReadOmrPrefix(mLocalOmrPrefix) != OT_ERROR_NONE || !IsValidOmrPrefix(mLocalOmrPrefix))
+    {
+        Ip6::NetworkPrefix randomOmrPrefix;
+
+        otLogNoteBr("no valid OMR prefix found in settings, generating new one");
+
+        if (randomOmrPrefix.GenerateRandomUla() != OT_ERROR_NONE)
+        {
+            otLogCritBr("failed to generate random OMR prefix");
+            ExitNow(error = OT_ERROR_FAILED);
+        }
+
+        mLocalOmrPrefix.Set(randomOmrPrefix);
+        IgnoreError(Get<Settings>().SaveOmrPrefix(mLocalOmrPrefix));
+    }
+
+exit:
+    return error;
+}
+
+otError RoutingManager::LoadOrGenerateRandomOnLinkPrefix(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    if (Get<Settings>().ReadOnLinkPrefix(mLocalOnLinkPrefix) != OT_ERROR_NONE ||
+        !IsValidOnLinkPrefix(mLocalOnLinkPrefix))
+    {
+        Ip6::NetworkPrefix randomOnLinkPrefix;
+
+        otLogNoteBr("no valid on-link prefix found in settings, generating new one");
+
+        if (randomOnLinkPrefix.GenerateRandomUla() != OT_ERROR_NONE)
+        {
+            otLogCritBr("failed to generate random on-link prefix");
+            ExitNow(error = OT_ERROR_FAILED);
+        }
+
+        randomOnLinkPrefix.m8[6] = 0;
+        randomOnLinkPrefix.m8[7] = 0;
+        mLocalOnLinkPrefix.Set(randomOnLinkPrefix);
+
+        IgnoreError(Get<Settings>().SaveOnLinkPrefix(mLocalOnLinkPrefix));
+    }
+
+exit:
+    return error;
+}
+
+void RoutingManager::Start(void)
+{
+    if (!mIsRunning)
+    {
+        StartRouterSolicitation();
+        mIsRunning = true;
```

Comment:
```suggestion
        mIsRunning = true;
        StartRouterSolicitation();
```

set `mIsRunning` as early as possible?